### PR TITLE
Code fixes and reverting changes

### PR DESCRIPTION
**Pull request by sudhacheran**
Code changes for the following
1. Reverted the cursor changes in StringInput and NumberInput Component
2. Fix for #146 
3. Adding hint to the StringInput and NumberInput components
4. Added fix for #152 from #156 

Tested the changes in MacBook - safari and chrome, IOS (iPad)-  Safari and Chrome and Android (Samsung phone) - Chrome

### CODE REVIEW

File: src/App.css
Comment (charliepnnl) [inline src/App.css:1]: I don't know why, but everything is rendered larger in this version of our code compared to what is at quality-install-tool.pnnl.gov. Compare the two screenshots below. The font size in neither of these versions seem to scale as the screen width changes - which also surprises me. I have not compared these with our code in main.
Screen Shot 2023-08-22 at 7 11 55 PM
Screen Shot 2023-08-22 at 7 12 06 PM

File: src/components/number_input.tsx [hunk 1]
```diff
@@ -70,16 +65,15 @@ const NumberInput: FC<NumberInputProps> = ({
...
                     onChange={event => {
-                        setCursor(event.target.selectionStart) // Set the cursor position as the selectionStart
+                        setLocalValue(event.target.value)
                         handleChange(event.target.value)
                     }}
-                    type="text"
-                    inputMode="numeric"
-                    value={value != null ? value : ''}
+                    type="number"
+                    value={localValue}
```
Comment (charliepnnl) [inline src/components/number_input.tsx:73]: `value` should be set with `value` not `localValue`.

File: src/components/number_input.tsx [hunk 2]
```diff
@@ -38,21 +39,15 @@ const NumberInput: FC<NumberInputProps> = ({
...
-        const input = ref.current
-        if (input) input.setSelectionRange(cursor, cursor)
-    }, [ref, cursor, value])
+    const [localValue, setLocalValue] = useState<string>(
+        value as unknown as string,
+    )
 
     const handleChange = (inputValue: string): any => {
-        const inputValueNum: number = parseInt(inputValue)
+        const inputValueNum: number = parseFloat(inputValue)
```
Comment (charliepnnl) [inline src/components/number_input.tsx:50]: We should send this inputValueNum to `updateValue`. In other words, we should be storing the number as a number in the document.

File: src/components/string_input.tsx [hunk 3]
```diff
@@ -32,19 +33,10 @@ const StringInput: FC<StringInputProps> = ({
...
-    const ref = useRef<HTMLInputElement>(null)
-
-    // Refresh after the first render and every time the component updates
-    useEffect(() => {
-        const input = ref.current
-        if (input) input.setSelectionRange(cursor, cursor)
-    }, [ref, cursor, value])
-
     const handleChange = (inputValue: string) => {
         if (typeof inputValue !== 'string') {
```
Comment (charliepnnl) [inline src/components/string_input.tsx:41]: Is it possible for inputValue to NOT be a string?

File: src/components/string_input.tsx [hunk 4]
```diff
@@ -63,15 +55,12 @@ const StringInput: FC<StringInputProps> = ({
...
-                    onChange={event => {
-                        setCursor(event.target.selectionStart) // Set the cursor position as the selectionStart
-                        handleChange(event.target.value)
-                    }}
+                    onChange={event => handleChange(event.target.value)}
                     type="text"
                     value={value || ''}
                     isInvalid={Boolean(error)}
                 />
+                <Form.Text>{hint}</Form.Text>
```
Comment (charliepnnl) [inline src/components/string_input.tsx:63]: Please check whether this still takes up vertical space when hint is empty. You may need to wrap this with conditional code:  `hint ? <Form.Text>{hint}</Form.Text> : null`

**Review by charliepnnl — APPROVED** [review]
Please merge this into main after addressing my comments and testing.

**Comment by charliepnnl** [comment]
Oh, and I believe we've used `rem` thoughout our CSS rather than `em`. If that is true, please use `rem` for consistency.

File: src/App.css
Comment (sudhacheran) [reply src/App.css:1]: In the local app screen, there seems a zoom applied (zoom icon in the address bar). In my system, I couldn't see this difference.  We could verify it once more before merging tomorrow.
 
Screenshot 2023-08-22 at 9 34 29 PM

File: src/components/number_input.tsx (hunk 1 above)
Comment (sudhacheran) [reply src/components/number_input.tsx:73]: with  value={value != null ? value : ''} works fine in chrome but did not work in safari. So changed to useState variable localValue. It works fine.

File: src/components/number_input.tsx (hunk 2 above)
Comment (sudhacheran) [reply src/components/number_input.tsx:50]: updateValue in number_input_wrapper.tsx uses parseFloat to upsertData which store value as float in the database. Changing updateValue with InputValueNum for some reason not working when I tested.

File: src/components/string_input.tsx (hunk 4 above)
Comment (sudhacheran) [reply src/components/string_input.tsx:63]: Yes, it takes vertical space. I have added a null check as suggested. It works fine now.

File: src/components/number_input.tsx (hunk 2 above)
Comment (charliepnnl) [reply src/components/number_input.tsx:50]: Ah. Ok. I didn't notice that value was being converted to a float in number_input_wrapper. That serves the same purpose.  I am resolved on this.

File: src/components/number_input.tsx (hunk 1 above)
Comment (charliepnnl) [reply src/components/number_input.tsx:73]: Ok. I see what is going on. If the input string is something like ".", then parseFloat() returns `NaN` which is what is stored in the Store and in the DB.  The "." is lost.

So I see two options:
1) Store the number input value as a string in the Store and in the DB. 
2) Store the number input value in local state as a string and call `updateValue()` only when `parseFloat()` is not NaN.

My preference is for the second option. So, very close to the code that you have now.
We can talk about this more.

File: src/components/string_input.tsx (hunk 3 above)
Comment (sudhacheran) [reply src/components/string_input.tsx:41]: Removed the typeof check for string

File: src/components/number_input.tsx (hunk 1 above)
Comment (sudhacheran) [reply src/components/number_input.tsx:79]: Have modified the as per discussion with Charlie.  Updated in numberInput -   the call updateValue() is made only when parseFloat() is not NaN. Additional check was implement to show error, when number not in the expected range gets stored in the database e.g., negative numbers when min is 0.

**Comment by sudhacheran** [comment]
Tested in Safari and Chrome form Mac and Android -chrome.

**Review by charliepnnl — APPROVED** [review]

**Pull request merged**